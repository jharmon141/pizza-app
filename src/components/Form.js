import React from 'react'
import { Link } from 'react-router-dom'
import { Col, Row } from 'react-bootstrap'

class Form extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            submitted: false,
            pickedToppings: [],
            toppingsMax: false ,
            maxNumberOfToppings: 0,
            total: 0
        }

        this.addTopping = this.addTopping.bind(this)
        this.removeTopping = this.removeTopping.bind(this)
        this.handleToppingChange = this.handleToppingChange.bind(this)
        this.updateTotal = this.updateTotal.bind(this)
        this.handleAddToCart = this.handleAddToCart.bind(this)
    }

    handleAddToCart() {
        let pizza = {}
        pizza.size = this.props.pizza.name
        pizza.toppings = this.state.pickedToppings
        pizza.total = this.state.total

        this.props.handleAddPizza(pizza)
        this.setState({submitted: true})
    }

    handleToppingChange(topping) {
        this.state.pickedToppings.includes(topping) ? this.removeTopping(topping) : this.addTopping(topping)
    }

    updateTotal(change, amount) {
        if (change === 'add') {
            let toppingPrice = Number(amount)
            let beginTotal = Number(this.state.total)
            let newTotal = beginTotal + toppingPrice
            this.setState({total: newTotal})
        }
        else if (change === 'subtract') {
            let toppingPrice = Number(amount)
            let beginTotal = Number(this.state.total)
            let newTotal = beginTotal - toppingPrice
            this.setState({total: newTotal})
        }
    }

    addTopping(topping) {
        if (!this.state.toppingsMax) {
            let newPickedToppings = this.state.pickedToppings.concat(topping)
            this.setState({pickedToppings: newPickedToppings})
            this.updateTotal('add', topping.price)
            if (newPickedToppings.length === this.props.pizza.maxToppings) {
                this.setState({toppingsMax: true})
            }
        } else {
            return
        }
    }

    removeTopping(topping) {
        let newToppings = this.state.pickedToppings.filter(each => {
            return each !== topping
        })
        this.setState({pickedToppings: newToppings})
        this.updateTotal('subtract', topping.price)
        if (newToppings.length < this.props.pizza.maxToppings) {
            this.setState({toppingsMax: false})
        }
    }

    componentDidMount() {
        let defaultToppings = []
        let sum = this.props.pizza.basePrice
        this.props.pizza.toppings.forEach(topping => {
            if (topping.defaultSelected) {
                defaultToppings.push(topping.topping)
                sum += topping.topping.price
            }
            this.addTopping(defaultToppings)
        })
        this.setState({total: sum})
        if (this.props.pizza.maxToppings !== null) {
            this.setState({maxNumberOfToppings: this.props.pizza.maxToppings})
        } else {
            this.setState({maxNumberOfToppings: 'Unlimited!'})
        }
    }

    render() {
        if (this.state.submitted) {
            return(
                <Row>
                    <Col xs={4} xsOffset={4}>
                        <h1>Pizza added!</h1>
                        <span
                            className="button"
                            onClick={this.props.resetForm}>
                            Add more pizza 
                        </span>
                        <br />
                        <br />
                        <Link className="cart" to='/cart'>
                            <span
                                className="button">
                                Go to cart
                            </span>
                        </Link>
                    </Col>
                </Row>
            )
        }

        else {
            return(
                <div>
                    <Row className="row">
                        <h2>Size: {this.props.pizza.name.toUpperCase()}</h2>
                        <h4>Base price: ${this.props.pizza.basePrice.toFixed(2)}</h4>
                        <h5>Max number of toppings: {this.state.maxNumberOfToppings}</h5>
                    </Row>
                    <Row>
                        <Col xs={4} xsOffset={2}>
                            <h4>Select your toppings:</h4>
                        </Col>
                        <Col xs={4} xsOffset={0}>
                            <form className="toppings">
                                {this.props.pizza.toppings.map(each => {
                                    if (this.state.toppingsMax && !this.state.pickedToppings.includes(each.topping)) {
                                        return(
                                            <div key={each.topping.name}>
                                                <label>
                                                    <input 
                                                        type="checkbox" 
                                                        onChange={() => this.handleToppingChange(each.topping)}
                                                        defaultChecked={each.defaultSelected}
                                                        name={each.topping.name} 
                                                        value={each.topping} disabled/> 
                                                    {each.topping.name} (${each.topping.price.toFixed(2)})
                                                </label>
                                                <br />
                                            </div>
                                        )
                                    }
                                    return(
                                        <div key={each.topping.name}>
                                            <label>
                                                <input 
                                                    type="checkbox" 
                                                    onChange={() => this.handleToppingChange(each.topping)}
                                                    defaultChecked={each.defaultSelected}
                                                    name={each.topping.name} 
                                                    value={each.topping}/> 
                                                {each.topping.name} (${each.topping.price.toFixed(2)})
                                            </label>
                                            <br />
                                        </div>
                                    )
                                })}
                            </form>
                        </Col>
                    </Row>
                    <Row>
                        <h4>Total: ${this.state.total.toFixed(2)}</h4>
                    </Row>
                    <Row>
                        <span
                            className="button"
                            onClick={this.handleAddToCart}>
                            Add to Cart
                        </span>
                        <br />
                        <br />
                        <span
                            className="button"
                            onClick={this.props.resetForm}>
                            Go back
                        </span>
                    </Row>
                </div>
            )
        }
    }
}

export default Form
