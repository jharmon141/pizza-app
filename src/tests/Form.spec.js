import React from 'react';
import Form from '../components/Form';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

describe('Form Component', () => {
   it('renders without crashing', () => {
      let pizza = {
         __typename: 'pizzaSize',
         basePrice: 9.89,
         maxToppings: 3,
         name: 'small',
         toppings: [
            {
               __typename: 'pizzaToppingConnection',
               defaultSelected: false,
               topping: {
                  __typename: 'topping',
                  name: 'pepperoni',
                  price: 0.4
               }
            }
         ]
      };

      const tree = renderer.create(
         <MemoryRouter> 
            <Form 
               pizza={pizza}
               handleRemovePizza={() => {}}
               handleAddPizza={() => {}}
            />
         </MemoryRouter>
      ).toJSON();
      expect(tree).toMatchSnapshot();
      setTimeout(() => done());
   })
})
