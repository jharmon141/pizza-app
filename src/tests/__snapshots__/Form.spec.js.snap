// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component renders without crashing 1`] = `
<div>
  <div
    className="row row"
  >
    <h2>
      Size: 
      SMALL
    </h2>
    <h4>
      Base price: $
      9.89
    </h4>
    <h5>
      Max number of toppings: 
      3
    </h5>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-4 col-xs-offset-2"
    >
      <h4>
        Select your toppings:
      </h4>
    </div>
    <div
      className="col-xs-4 col-xs-offset-0"
    >
      <form
        className="toppings"
      >
        <div>
          <label>
            <input
              defaultChecked={false}
              name="pepperoni"
              onChange={[Function]}
              type="checkbox"
              value={
                Object {
                  "__typename": "topping",
                  "name": "pepperoni",
                  "price": 0.4,
                }
              }
            />
            pepperoni
             ($
            0.40
            )
          </label>
          <br />
        </div>
      </form>
    </div>
  </div>
  <div
    className="row"
  >
    <label
      className="quantityLabel"
    >
      Quantity:
      <select
        className="quantitySelect"
        onChange={[Function]}
        value={1}
      >
        <option
          value={1}
        >
          1
        </option>
        <option
          value={2}
        >
          2
        </option>
        <option
          value={3}
        >
          3
        </option>
        <option
          value={4}
        >
          4
        </option>
        <option
          value={5}
        >
          5
        </option>
        <option
          value={6}
        >
          6
        </option>
        <option
          value={7}
        >
          7
        </option>
        <option
          value={8}
        >
          8
        </option>
        <option
          value={9}
        >
          9
        </option>
        <option
          value={10}
        >
          10
        </option>
      </select>
    </label>
  </div>
  <div
    className="row"
  >
    <h4>
      Total: $
      9.89
    </h4>
  </div>
  <div
    className="row"
  >
    <span
      className="button"
      onClick={[Function]}
    >
      Add to Cart
    </span>
    <br />
    <br />
    <span
      className="button"
      onClick={undefined}
    >
      Go back
    </span>
  </div>
</div>
`;
